### ISSUE TITLE
tf.nn.embedding_lookup  poor performance

### ISSUE THREAD

**Issue by chengdianxuezi** [tensorflow/tensorflow#9777]
In my programe, I use  f.nn.embedding_lookup as follow:

embedding = tf.get_variable("embedding", [200000, 128], tf.float32, initializer=tf.        random_normal_initializer(stddev=0.1), trainable=True,partitioner=tf.fixed_size_partitioner(10))

word_embedding = tf.nn.embedding_lookup(embedding, query_tensor)
while(1):
    sess.run(word_embedding )

when I start 30 worker，each worker qps is 3000.
but when I test tf.nn.embedding_lookup_sparse, each worker qps is 110000

how to solve this problem,any suggestion welcome

**Comment by chengdianxuezi** [comment]
query_tensor shape is [1000, 20]

**Comment by suiyuan2009** [comment]
try [embedding_lookup_sparse_with_distributed_aggregation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/embedding_ops.py#L566)

**Comment by wsnooker** [comment]
Same problem i got. What is the inner difference implementation in `embedding_lookup_sparse ` and `embedding_lookup `  ? In my view, it is just index lookup and some combination added.

**Comment by suiyuan2009** [comment]
`embedding_lookup_sparse` invokes `embedding_lookup `, then does some combination. The implementation is different if `sp_weights` is passed or not, which will affect parameter server's load. If `sp_weights` is passed, ps's load is higher.

**Comment by chengdianxuezi** [comment]
@suiyuan2009  I guess the combination of embedding_lookup_sparse is done in the ps, so the result of embedding_lookup_sparse  is very small ,the shape about [1000,128], but the result of embedding_lookup is a dense tensor,and the shape is [1000,20,128].  the bottleneck is ps network

**Comment by suiyuan2009** [comment]
@chengdianxuezi , the official api does combination on the machine running `embedding_lookup_sparse` op. There is a modified api [embedding_lookup_sparse_with_distributed_aggregation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/embedding_ops.py#L566) which does combination on ps side.

**Comment by chengdianxuezi** [comment]
@suiyuan2009 I use tf.nn.embedding_lookup_sparse and tf.nn.embedding_lookup in the same way,why the  speed gap is great。the speed of tf.nn.embedding_lookup should be fast then tf.nn.embedding_lookup_sparse

**Comment by suiyuan2009** [comment]
@chengdianxuezi , if `sp_weights` is not passed, duplicates in `sp_ids` will be removed before invoking `embedding_lookup`.

**Comment by wsnooker** [comment]
@suiyuan2009 
I tested the function  `tf.contrib.layers.safe_embedding_lookup_sparse`, which is much more slower than `tf.nn.embedding_lookup_sparse`.  I am not sure whether the time wasted in padding additional values.

**Comment by suiyuan2009** [comment]
@wsnooker , try `embedding_lookup_sparse_with_distributed_aggregation`, also I have not tested this api...

**Comment by chengdianxuezi** [comment]
@suiyuan2009 Thank you very much,  use the new api tf.contrib.layers.embedding_lookup_unique, the performance improved, best wish to you

**Comment by suiyuan2009** [comment]
@chengdianxuezi , can you help to test the performance of `embedding_lookup_sparse_with_distributed_aggregation`, I write the api but I don't know its performance...

**Comment by wsnooker** [comment]
@suiyuan2009 In the TF distribution, all the variables stored in ps, and the worker has to communicate with ps in the forward and backward. This is a wasteful process if big batch data presented.

**Comment by suiyuan2009** [comment]
@wsnooker , putting some computation on ps will reduce network traffic and improve parallelism. there is a [paper](https://arxiv.org/abs/1606.08495) at CIKM 2016, they developed a distributed word2vec training system that shares the same thought.

**Comment by wsnooker** [comment]
@suiyuan2009  What kind of computation will be put on ps in TF framework ？ As far as I know, the `embedding_lookup` op finished on ps. What about `conv` / `rnn` and some others ?  There is little material about this in TF web.

**Comment by suiyuan2009** [comment]
@wsnooker , I rewrited embedding_lookup mechanism and put some computation on ps.

**Comment by skye** [comment]
Closing due to lack of activity. Please reopen if this is still an issue.

**Comment by chunyang-wen** [comment]
@suiyuan2009 Why don't you check the performance first before check in your code? It makes no sense. BTW, I have drew the timeline for ***_distributed_aggregation, sure, part of the computation has been moved to ps, but the overall performance does not improve.

**Comment by formath** [comment]
The performance of `embedding_lookup_sparse` for million of rows is very poor. More precisely, the bottleneck is on the network between worker and ps. As a ultimate solution, `RowSparsePull` and `RowSparsePush` are needed. Anyone want to contribute? @suiyuan2009

**Comment by hkxIron** [comment]
@chengdianxuezi ,embedding_lookup_unique now can't by found in tensorflow ,do you still use it?
